src/adapter/src/coord/sequencer/inner/create_view.rs

```diff
@@ -9,81 +9,72 @@
+    StageResult, Staged,
 };
 use crate::error::AdapterError;
 use crate::optimize::{self, Optimize};
 use crate::session::Session;
 use crate::{catalog, AdapterNotice, ExecuteContext};
 
+impl Staged for CreateViewStage {
+    fn validity(&mut self) -> &mut PlanValidity {
+        match self {
+            Self::Optimize(stage) => &mut stage.validity,
+            Self::Finish(stage) => &mut stage.validity,
+        }
+    }
+
+    async fn stage(
+        self,
+        coord: &mut Coordinator,
+        session: &Session,
+    ) -> Result<StageResult<Box<Self>>, AdapterError> {
+        match self {
+            CreateViewStage::Optimize(stage) => coord.create_view_optimize(stage).await,
+            CreateViewStage::Finish(stage) => coord.create_view_finish(session, stage).await,
+        }
+    }
+
+    fn message(self, ctx: ExecuteContext, span: Span) -> Message {
+        Message::CreateViewStageReady {
+            ctx,
+            span,
+            stage: self,
+        }
+    }
+}
+
 impl Coordinator {
     pub(crate) async fn sequence_create_view(
         &mut self,
         ctx: ExecuteContext,
         plan: plan::CreateViewPlan,
         resolved_ids: ResolvedIds,
     ) {
-        self.sequence_create_view_stage(
-            ctx,
-            CreateViewStage::Validate(CreateViewValidate { plan, resolved_ids }),
-            OpenTelemetryContext::obtain(),
-        )
-        .await;
-    }
-
-    /// Processes as many `create view` stages as possible.
-    #[instrument(skip_all)]
-    pub(crate) async fn sequence_create_view_stage(
-        &mut self,
-        mut ctx: ExecuteContext,
-        mut stage: CreateViewStage,
-        otel_ctx: OpenTelemetryContext,
-    ) {
-        use CreateViewStage::*;
-
-        // Process the current stage and allow for processing the next.
-        loop {
-            // Always verify plan validity. This is cheap, and prevents programming errors
-            // if we move any stages off thread.
-            if let Some(validity) = stage.validity() {
-                return_if_err!(validity.check(self.catalog()), ctx);
-            }
-
-            (ctx, stage) = match stage {
-                Validate(stage) => {
-                    let next = return_if_err!(self.create_view_validate(ctx.session(), stage), ctx);
-                    (ctx, CreateViewStage::Optimize(next))
-                }
-                Optimize(stage) => {
-                    self.create_view_optimize(ctx, stage, otel_ctx).await;
-                    return;
-                }
-                Finish(stage) => {
-                    let result = self.create_view_finish(&mut ctx, stage).await;
-                    ctx.retire(result);
-                    return;
-                }
-            }
-        }
+        let stage = return_if_err!(
+            self.create_view_validate(ctx.session(), plan, resolved_ids),
+            ctx
+        );
```

Comment:
Maybe we can get rid of `create_view_validate` and move its logic into here?

Reply:
I did that at one point, but that fn has multiple error returns, so it turned out to be more convenient as a function where `?` works, then a single `return_if_err` in the outer fn.